third_party/xla_client/metrics_analysis.cc

```diff
@@ -0,0 +1,204 @@
+      return {
+          Analysis::Symptom::kMetricTooSlow,
+          absl::StrFormat("%s too slow: longest instance took %f seconds. "
+                          "Please open an GitHub issue with the graph dump for "
+                          "our team to optimize.",
+                          metric_name_, max_metric_time / 1e9),
+      };
+    }
+    return {Analysis::Symptom::kNormal};
+  }
+
+ private:
+  std::string metric_name_;
+  long threshold_nsec_;
+};
+
+class XrtMetricFrequency : public Analyzer {
+ public:
+  XrtMetricFrequency(std::map<std::string, float> metric_name_thresholds,
+                     int run_every_n)
+      : metric_name_thresholds_(std::move(metric_name_thresholds)),
+        run_every_n_(run_every_n),
+        counter_(0) {}
+
+  Analysis Run() override {
+    // XRT GetMetrics call is relatively expensive.
+    if (++counter_ % run_every_n_ != 0) {
+      return {Analysis::Symptom::kNormal};
+    }
+    counter_ = 0;
+    CounterData* step = GetCounter("MarkStep");
+    if (!step) {
+      return {Analysis::Symptom::kNormal};
+    }
+
+    std::stringstream ss;
+    xla::int64 step_count = step->Value();
+    auto xrt_metrics = ComputationClient::Get()->GetMetrics();
+    for (auto const& kv : metric_name_thresholds_) {
+      auto it = xrt_metrics.find(kv.first);
+      if (it == xrt_metrics.end()) {
+        continue;
+      }
+      xla::Metric xrt_metric = it->second;
+      std::ldiv_t res;
+      if (xrt_metric.int64_value) {
+        int64 metric_count = *xrt_metric.int64_value;
+        res = std::div(metric_count, step_count);
+      } else if (xrt_metric.percentile) {
+        size_t metric_count = (*xrt_metric.percentile).total_samples;
+        res = std::div(metric_count, step_count);
+      } else {
+        continue;
+      }
+      if (DivToFloat(res, step_count) > kv.second) {
+        ss << kv.first << " (" << DivToFloat(res, step_count) * step_count
+           << " counts), ";
+      }
+    }
+
+    std::string repr = ss.str();
+    if (!repr.empty()) {
+      return {
+          Analysis::Symptom::kMetricTooFrequent,
+          absl::StrFormat("Following metrics too frequent: %sduring %zu steps. "
+                          "Note: XRT metrics follow the lifecycle of the TPU "
+                          "so you may need "
+                          "to restart the TPU for fresh metrics.",
+                          repr, step_count),
+      };
+    }
+    return {Analysis::Symptom::kNormal};
+  }
+
+ private:
+  std::map<std::string, float> metric_name_thresholds_;
+  bool is_xrt_metric_;
+  int run_every_n_;
+  int counter_;
+};
+
+class UnloweredOp : public Analyzer {
+ public:
+  Analysis Run() override {
+    std::stringstream ss;
+    MetricsArena* arena = MetricsArena::Get();
+    arena->ForEachCounter([&ss](const std::string& name, CounterData* data) {
+      if (absl::StrContains(name, "aten::") &&
+          name != "aten::_local_scalar_dense") {
```

Comment:
Should we also have a check for `aten::_local_scalar_dense` and `earlySync` and warn if it is being used too often?

Reply:
The detection of `aten::_local_scalar_dense` too frequent ops should be detected by `MetricFrequency("TransferFromServerTime")` analyzer.